### PR TITLE
[WIP] Feature/attendances

### PR DESCRIPTION
**Pull request by philipflohr**
This implements the attendance methods.

Still missing: Mock tests, in detail raw api tests

This PR needs #11 to be merged first

### CODE REVIEW

**Comment by philipflohr** [comment]
@klamann I'll get some time to finally integrate my changes to personio-py. Mock tests and tests using the authenticated API are available. The Sphinx build currently fails due to a dependency issue (mistune 0.8.4 / 2.0.0).

**Comment by klamann** [comment]
Hey @philipflohr, glad to hear it! I'm in the middle of the refactoring for personio-py 1.0, but if we merge this soon, I can integrate it into the refactored codebase.

To resolve the sphinx dependency issue, please add `mistune = "0.8.4"` to requirements-dev.txt

**Comment by philipflohr** [comment]
@klamann Pinning the package did not resolve the issue. From a very quick look at the logfile I guess that Sphinx has the same dependency issue

**Comment by philipflohr** [comment]
Hi @klamann 

I had closer look at the docs-verify action. In my opinion this is the line where pinning the mistune package needs to be done
`pre-build-command: "pip install sphinx_rtd_theme~=0.4.3 recommonmark~=0.6.0 m2r~=0.2.1"`

**Comment by philipflohr** [comment]
@klamann Could you have a look if the docs-verify check is fixed if you install the correct mistune package in the prebuild command?

**Comment by klamann** [comment]
I solved this in #24 with this set of dependencies:

```
Sphinx = "^4.3.2"
sphinx-rtd-theme = "^1.0.0"
recommonmark = "^0.7.1"
docutils = "0.16"
mistune = "0.8.4"
m2r = "^0.2.1"
```

the order of the dependencies should not matter, but I think you're missing `docutils = "0.16"` (I think there was an issue with the latest version, can't remember what it was though).

Also, you can change the `pre-build-command` in `docs-pr.yml` to `pip install -r requirements-dev.txt` if you want. All of this is changing in #24 anyways.

File: src/personio_py/models.py [hunk 1]
```diff
@@ -665,13 +665,49 @@ def to_dict(self, nested=False) -> Dict[str, Any]:
...
+        if patch_existing_attendance:
+            if self.id_ is None:
+                raise ValueError("An attendance id is required")
+            body_dict = {}
+            if self.date is not None:
+                body_dict['date'] = self.date.strftime("%Y-%m-%d")
+            if self.start_time is not None:
+                body_dict['start_time'] = str(self.start_time)[:-3]
+            if self.end_time is not None:
+                body_dict['end_time'] = str(self.end_time)[:-3]
```
Comment (klamann) [inline src/personio_py/models.py:695]: please use `DurationFieldMapping.serialize` for start_time and end_time

File: src/personio_py/models.py [hunk 2]
```diff
@@ -665,13 +665,49 @@ def to_dict(self, nested=False) -> Dict[str, Any]:
...
+            if self.comment is not None:
+                body_dict['comment'] = self.comment
+            return body_dict
+        else:
+            return \
+                {
+                    "employee": self.employee_id,
+                    "date": self.date.strftime("%Y-%m-%d"),
+                    "start_time": self.start_time,
+                    "end_time": self.end_time,
```
Comment (klamann) [inline src/personio_py/models.py:707]: does this work? I think we need to serialize these values too

File: tests/test_mock_api_attendances.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+def mock_attendances():
+    # mock the get absences endpoint (with different array offsets)
+    responses.add(
+        responses.GET, re.compile('https://api.personio.de/v1/company/attendances?.*'),
+        status=200, json=json_dict_attendance_rms, adding_headers={'Authorization': 'Bearer foo'})
+    
+def mock_create_attendance():
+    responses.add(
+        responses.POST,  'https://api.personio.de/v1/company/attendances',
```
Comment (klamann) [inline tests/test_mock_api_attendances.py:84]: can you enable pre-commit hooks, or check your code with `flake8 --max-line-length=100 --ignore=F401,W504 src`? There are some minor code style issues in this PR, like the double whitespace here.

**Review by klamann — COMMENTED** [review]
The implementation looks good to me! I have marked a few code style nitpicks, but other than that I think we can merge this.

If you want to release a new version, please update `version.py` and add a Changelog entry (see [readme#pypi-release](https://github.com/at-gmbh/personio-py#pypi-release))

**Pull request merged**